### PR TITLE
GitRepo Initial Checkout when DisablePolling is True

### PR DESCRIPTION
**Pull request by Tommy12789**
Refers to #2432

implemented an initial checkout for the gitrepos applied without polling and commit update while force updating 

./integrationtests/gitjob/controller/controller_test.go

### CODE REVIEW

File: internal/cmd/controller/gitops/reconciler/gitjob_controller.go [hunk 1]
```diff
@@ -103,9 +104,27 @@ func (r *GitJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
+	if errors.IsNotFound(err) {
+		if gitRepo.Status.Commit != "" {
+			if err := r.createJob(ctx, &gitRepo); err != nil {
+				return ctrl.Result{}, fmt.Errorf("error creating git job: %v", err)
+			}
+		} else {
+			if gitRepo.Status.GitJobStatus == status.FailedStatus.String() {
+				return ctrl.Result{}, fmt.Errorf("error creating git job: %v", err)
+			}
+
```
Comment (0xavi0) [inline internal/cmd/controller/gitops/reconciler/gitjob_controller.go:116]: Why are we checking  for the `GitJobStatus` if the job was not found? 
Also... the log message seems wrong...

**Review by 0xavi0 — COMMENTED** [review]
As discussed offline, there is a bug when doing a "Force Update" because the contents are re-deployed but the `commit` is not updated.

That works because the completed job remains in the cluster and the 
```go
	} else if gitRepo.Status.Commit != "" {
		if err = r.deleteJobIfNeeded(ctx, &gitRepo, &job); err != nil {
			return ctrl.Result{}, fmt.Errorf("error deleting git job: %v", err)
		}
	}
```
code is executed (which recreates the job and fetches the latest from the repository).

But this part:
```go
                        fetcher := git.NewFetcher()
			commit, err := fetcher.LatestCommit(ctx, &gitRepo, r.Client)
			if err != nil {
				return ctrl.Result{}, fmt.Errorf("error fetching commit: %v", err)
			}
			gitRepo.Status.Commit = commit
			logger.V(1).Info("Updating GitRepo status", "commit", gitRepo.Status.Commit)
			if err := r.Status().Update(ctx, &gitRepo); err != nil {
				return ctrl.Result{}, fmt.Errorf("error updating git repo status: %v", err)
			}
			return ctrl.Result{}, nil
``` 
is no longer executed and the `commit` is not updated in the `gitrepo`.

It is not executed because it deletes the job and then it executes this branch:
```go
     if errors.IsNotFound(err) {
		if gitRepo.Status.Commit != "" {
			if err := r.createJob(ctx, &gitRepo); err != nil {
				return ctrl.Result{}, fmt.Errorf("error creating git job: %v", err)
			}
		}
```
as `gitRepo.Status.Commit` is no longer `""` like it was for the first time.

File: internal/cmd/controller/gitops/reconciler/gitjob_controller.go (hunk 1 above)
Comment (Tommy12789) [reply internal/cmd/controller/gitops/reconciler/gitjob_controller.go]: Of course, I removed these lines

**Review by 0xavi0 — COMMENTED** [review]
I guess it would be cool to also test the case that had the bug before, but I guess that requires an `e2e` test because you need to create a new commit and these tests are relying on the `fleet` repo in github. 
Up to you if you want to experiment and create that.


Aside from the code simplification and the lines that belong to other PRs... LGTM

File: integrationtests/gitjob/controller/controller_test.go [hunk 2]
```diff
@@ -217,6 +217,28 @@ var _ = Describe("GitJob controller", func() {
...
+		It("should update the commit from the actual repo", func() {
+			gitRepo = createGitRepoWithDisablePolling(gitRepoName)
+			Expect(k8sClient.Create(ctx, &gitRepo)).To(Succeed())
+
+			By("verifying the commit is updated")
+			Eventually(func() string {
+				Expect(k8sClient.Get(ctx, types.NamespacedName{Name: gitRepoName, Namespace: gitRepoNamespace}, &gitRepo)).To(Succeed())
+				return gitRepo.Status.Commit
+			}, "30s", "1s").ShouldNot(Equal(""))
+		})
```
Comment (0xavi0) [inline integrationtests/gitjob/controller/controller_test.go:240]: I think this test could also check that the job is eventually created, like other tests in this same file.

File: internal/cmd/controller/gitops/reconciler/gitjob_controller.go [hunk 3]
```diff
@@ -103,9 +104,29 @@ func (r *GitJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
+			}
+		} else {
+			if err := r.updateCommit(ctx, &gitRepo); err != nil {
+				if errors.IsConflict(err) {
+					logger.V(1).Info("conflict updating commit, retrying", "message", err)
+					return ctrl.Result{Requeue: true}, nil // just retry, but don't show an error
+				}
+				return ctrl.Result{}, fmt.Errorf("error updating commit: %v", err)
+			}
+			return ctrl.Result{}, nil
```
[the remark this answers is not in the document]
Comment (Tommy12789) [reply internal/cmd/controller/gitops/reconciler/gitjob_controller.go:129]: Of course I"ll change that

File: integrationtests/gitjob/controller/controller_test.go (hunk 2 above)
Comment (Tommy12789) [reply integrationtests/gitjob/controller/controller_test.go:240]: I"ll do it !

File: internal/cmd/controller/gitops/reconciler/gitjob_controller.go [hunk 4]
```diff
@@ -172,6 +205,8 @@ func (r *GitJobReconciler) updateStatus(ctx context.Context, gitRepo *v1alpha1.G
 		}
 
 		currentGitRepo.Status.GitJobStatus = result.Status.String()
+		currentGitRepo.Status.ObservedGeneration = gitRepo.Generation
+		currentGitRepo.Status.UpdateGeneration = gitRepo.Status.UpdateGeneration
```
[the remark this answers is not in the document]
Comment (Tommy12789) [reply internal/cmd/controller/gitops/reconciler/gitjob_controller.go:209]: Yes I had some conflicts so I added

File: internal/cmd/controller/gitops/reconciler/gitjob_controller.go [hunk 5]
```diff
@@ -103,9 +104,29 @@ func (r *GitJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 		return ctrl.Result{}, fmt.Errorf("error retrieving git job: %v", err)
 	}
 
-	if errors.IsNotFound(err) && gitRepo.Status.Commit != "" {
-		if err := r.createJob(ctx, &gitRepo); err != nil {
-			return ctrl.Result{}, fmt.Errorf("error creating git job: %v", err)
+	if errors.IsNotFound(err) {
+		if gitRepo.Status.Commit != "" {
```
Comment (manno) [inline internal/cmd/controller/gitops/reconciler/gitjob_controller.go:108]: Is there an easier way to express this logic which uses not as many nested ifs?

Maybe 

```
if commit == "" || disablePolling { updateCommit }
if commit != "" { createJob }
```

File: pkg/git/fetch.go [hunk 6]
```diff
@@ -18,6 +18,10 @@ const (
 
 type Fetch struct{}
 
+func NewFetcher() *Fetch {
```
Comment (manno) [inline pkg/git/fetch.go:21]: I would have expected the name of the struct.

```suggestion
func NewFetch() *Fetch {
```

File: e2e/single-cluster/gitrepo_polling_disabled_test.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+		_, err := k.Delete("gitrepo", gitrepoName)
+		Expect(err).ToNot(HaveOccurred())
+
+		Eventually(func() string {
+			out, _ := k.Get("bundledeployments", "-A")
+			return out
+		}).ShouldNot(ContainSubstring(gitrepoName))
+
+		out, err := k.Namespace("cattle-fleet-system").Logs(
```
Comment (0xavi0) [inline e2e/single-cluster/gitrepo_polling_disabled_test.go:63]: Nit: This could be flagged as a failure if any previous test failed.

**Review by 0xavi0 — APPROVED** [review]
LGTM
Thanks for adding the extra `e2e` test

**Review by 0xavi0 — APPROVED** [review]

**Pull request merged**